Dialyzer fixes.

cmeiklejohn opened the pull request:
Address a series of dialyzer bugs.  Some, but not all.  @seancribbs

cmeiklejohn committed:
Normalize the return value.

Ensure this function only ever returns the 'ok' atom for all request types.

cmeiklejohn committed:
Ignore return from supervisor calls.

Ignore the return values from the terminate_child and delete_child calls to supervisor.

cmeiklejohn committed:
Ignore the result of sending on the socket.

Since flow control is coded into the application, ignore the result of directly sending on the socket.

cmeiklejohn committed:
Ignore result of timer calls.

When canceling a timer, ignore the value of the cancellation.

cmeiklejohn committed:
Add ctags to the ignore file.

cmeiklejohn committed:
Ignore the result of multicall.

cmeiklejohn committed:
Ignore result of list comprehensions.

cmeiklejohn committed:
Ignore the result of the supervisor return.

cmeiklejohn committed:
Fix incorrect error handling.

Fix incorrect handling of the error response.

cmeiklejohn committed:
Fix incorrect types.

Adjust the type signature to properly reflect successful calls.

cmeiklejohn committed:
Remove a bunch of incorrect signatures.

Remove a bunch of incorrect signatures, with the intent to add the signatures back with typer once we are dialyzer clean.

cmeiklejohn committed:
Ignore return from supervisor calls.

cmeiklejohn committed:
Remove error tuple handling.

Remove handling for a return which is no longer possible to get based on the calling function.

cmeiklejohn committed:
Ignore return from the register function.

cmeiklejohn committed:
Refactor stats.

Refactor stats to crash if it is unable to register a particular stat, and make use of list comprehensions more efficient.

seancribbs commented on src/riak_core_cluster_mgr.erl:
```diff
@@ -108,7 +108,6 @@
 %%% API
 %%%===================================================================
 %% @doc start the Cluster Manager
--spec(start_link() -> ok).
```
I realize these specs are wrong, but is there a good reason to remove them from API functions?

cmeiklejohn replied:
I put in the commit description that the goals was to re-add the signatures with typer once dialyzer was green.

seancribbs replied:
Cool, thanks.

seancribbs commented on src/riak_core_service_mgr.erl:
```diff
@@ -117,14 +117,12 @@ start_link({IP,Port}) when is_integer(Port), Port >= 0 ->
 %% for taking ownership of the socket via
 %% Transport:controlling_process(Socket, Pid).  Only the strategy of
 %% `round_robin' is supported; it's arg is ignored.
--spec(register_service(hostspec(), service_scheduler_strategy()) -> ok).
```
Ditto comment re: removal of specs from API functions

seancribbs commented on src/riak_repl2_fscoordinator.erl:
```diff
@@ -314,8 +314,8 @@ handle_cast(start_fullsync,  State) ->
-    [erlang:cancel_timer(Tref) || {_, {_, Tref}} <- State#state.whereis_waiting],
-    [begin
+    _ = [erlang:cancel_timer(Tref) || {_, {_, Tref}} <- State#state.whereis_waiting],
+    _ = [begin
```
I'd make a minor note to extract this into another function for clarity.

seancribbs commented on src/riak_repl2_ip.erl:
```diff
@@ -11,7 +11,6 @@
 %% @doc Given the result of inet:getifaddrs() and an IP a client has
 %%      connected to, attempt to determine the appropriate subnet mask.  If
 %%      the IP the client connected to cannot be found, undefined is returned.
--spec determine_netmask(Ifaddrs :: [{atom(), any()}], SeekIP :: string() | {integer(), integer(), integer(), integer()}) -> 'undefined' | binary().
```
Again, API function. Might be better to simply use the proper types, e.g. `inet:ip_address()`

seancribbs commented:
Now that the shoe is on the other foot... I think it would be more helpful to put the warnings that were removed in the extended commit message. Then I can see that your change actually addressed the original warning. On the whole though, this is a vast improvement, I count 89 warnings removed.

cmeiklejohn committed:
Store aggregated references in the state.

Even if we don't use them directly, store them incase we want to
demonitor the references at a future time.

cmeiklejohn committed:
Enforce the metric can be registered.

Resolves:

riak_repl_stats.erl:167: Expression produces a value of type 'ok' | {'error',_,'unsupported_metric_type'}, but this value is unmatched

cmeiklejohn committed:
Discard the number as it's not used.

Resolves:

riak_repl_migration.erl:109: Expression produces a value of type {'ok',number()}, but this value is unmatched

cmeiklejohn committed:
Ignore return value.

Ignore the return value which can be either an empty list or a proplist.

Resolves:

riak_repl_ring_handler.erl:29: Expression produces a value of type [{'disabled',[any()]} | {'enabled',[any()]}], but this value is unmatched
riak_repl_ring_handler.erl:40: Expression produces a value of type [{'disabled',[any()]} | {'enabled',[any()]}], but this value is unmatched
riak_repl_ring_handler.erl:156: Expression produces a value of type [{'disabled',[any()]} | {'enabled',[any()]} | {'started',[any()]} | {'stopped',[any()]}], but this value is unmatched

cmeiklejohn committed:
Ignore return value.

Resolves:

riak_repl_listener_sup.erl:43: Expression produces a value of type [any()], but this value is unmatched

cmeiklejohn committed:
Correct socket type to support SSL.

Resolves:

riak_repl_tcp_server.erl:368: Record construction #state{sitename::'undefined' | string(),socket::{'sslsocket',{_,_},_},transport::'ranch_ssl',listener::'undefined' | pid(),client::'undefined' | tuple(),q::'undefined' | pid(),work_dir::'undefined' | string(),my_pi::'undefined' | #peer_info{riak_version::'undefined' | string(),repl_version::'undefined' | string(),ring::'undefined' | tuple()},their_pi::'undefined' | #peer_info{riak_version::'undefined' | string(),repl_version::'undefined' | string(),ring::'undefined' | tuple()},fullsync_worker::'undefined' | pid(),fullsync_strategy::atom(),election_timeout::'undefined' | reference(),keepalive_time::'undefined' | integer()} violates the declared type of field socket::'undefined' | port()

cmeiklejohn committed:
Normalize the return value.

Normalize the return value from the leader_change function, which is not
currently matched on at all.

Resolves:

riak_repl_leader.erl:489: Expression produces a value of type ['ok' | {'error','not_found' | 'restarting' | 'running' | 'simple_one_for_one'}], but this value is unmatched
riak_repl_leader.erl:428: Expression produces a value of type 'ok' | ['ok' | {'error','not_found' | 'restarting' | 'running' | 'simple_one_for_one'}], but this value is unmatched
riak_repl_leader.erl:357: Expression produces a value of type 'ok' | ['ok' | 'true' | {'error','not_found' | 'restarting' | 'running' | 'simple_one_for_one'}], but this value is unmatched

cmeiklejohn committed:
Ignore the result of the list comprehension.

Resolves:

riak_repl_leader.erl:437: Expression produces a value of type ['true'], but this value is unmatched

cmeiklejohn committed:
Ignore result of list comprehension.

Resolves:

riak_repl_leader.erl:491: Expression produces a value of type ['ok' | {'error','not_found' | 'restarting' | 'running' | 'simple_one_for_one'}], but this value is unmatched
riak_repl_leader.erl:497: Expression produces a value of type [any()], but this value is unmatched

cmeiklejohn committed:
Removes unknown type message for erlang:now().

cmeiklejohn committed:
Normalize return value.

Normalize the return value of this function, who's result is already
unmatched.

riak_repl_leader.erl:424: Expression produces a value of type ['ok' | {'error','not_found' | 'restarting' | 'running' | 'simple_one_for_one'}], but this value is unmatched
riak_repl_leader.erl:429: Expression produces a value of type 'ok' | ['ok' | {'error','not_found' | 'restarting' | 'running' | 'simple_one_for_one'}], but this value is unmatched

cmeiklejohn committed:
Fix type signature to resolve unknown type error.

cmeiklejohn committed:
Ignore the result of the list comprehension.

riak_repl_leader.erl:222: Expression produces a value of type [{'repl',_}], but this value is unmatched

cmeiklejohn committed:
Correct bad record typing.

Resolves:

riak_repl_leader.erl:159: The pattern {MRef, _, _} can never match the type 'false' | {reference(),pid()}
riak_repl_leader.erl:169: The pattern [{_Mref, Pid, 'send'} | _] can never match the type [{reference(),pid()}]
riak_repl_leader.erl:225: Record construction #state{i_am_leader::boolean(),leader_node::atom(),leader_mref::'undefined' | reference(),candidates::[atom()],workers::[atom()],receivers::[{_,pid(),'drop' | 'send'},...],check_tref::'undefined' | timer:tref(),elected_mbox_size::integer(),lastpoll::{non_neg_integer(),non_neg_integer(),non_neg_integer()}} violates the declared type of field receivers::[{reference(),pid()}]
riak_repl_leader.erl:227: The pattern [{_Mref, P, 'send'} | _] can never match the type [{reference(),pid()}]

cmeiklejohn committed:
Ignore the result of the os:cmd call.

Resolves:

riak_repl_keylist_server.erl:591: Expression produces a value of type string(), but this value is unmatched

cmeiklejohn committed:
Normalize return of schedule_fullsync.

This is fine to do because no callers of this function ever use the
reference which is returned from the send_after.

Resolves:

riak_repl_keylist_server.erl:138: Expression produces a value of type 'ok' | reference(), but this value is unmatched

seancribbs commented:
:+1: so far. Are you going to continue eliminating warnings or do we want this to merge?

cmeiklejohn commented:
Going to merge and start a new PR moving forward, does that work?

seancribbs commented:
Works for me.
